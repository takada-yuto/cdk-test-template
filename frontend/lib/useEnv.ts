import useSWR from "swr"

interface Env {
  apigatewayUrl: string | undefined
}

const fetcher = async () => {
  const res = await fetch("/env.json")
  return res.json()
}

export default function useEnv() {
  const { data: env } = useSWR<Env>("/env.json", fetcher, {
    revalidateOnFocus: false,
    revalidateOnReconnect: false,
  })

  if (!env) {
    return {
      env: Object.freeze({
        apigatewayUrl: process.env.NEXT_PUBLIC_API_GATEWAY_URL,
      }) as Env,
    }
  }

  return { env: Object.freeze(env) }
}
