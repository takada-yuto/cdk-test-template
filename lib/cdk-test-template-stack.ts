import * as cdk from "aws-cdk-lib"
import {
  CloudFrontWebDistribution,
  FunctionCode,
  FunctionEventType,
  FunctionRuntime,
  OriginAccessIdentity,
} from "aws-cdk-lib/aws-cloudfront"
import {
  AnyPrincipal,
  Effect,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam"
import { Code, Runtime } from "aws-cdk-lib/aws-lambda"
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs"
import { BlockPublicAccess, Bucket, ObjectOwnership } from "aws-cdk-lib/aws-s3"
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment"
import { Construct } from "constructs"
import { readFileSync } from "fs"
import path = require("path")
// import * as sqs from 'aws-cdk-lib/aws-sqs';

// AWS CDKを使用してCloudFormationスタックを定義するクラス
export class CdkTestTemplateStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    // CloudFrontのログを保存するためのS3バケットを作成
    const cdkTemplateLogBucket = new Bucket(this, "CdkTemplateLogBucket", {
      objectOwnership: ObjectOwnership.BUCKET_OWNER_PREFERRED,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    })
    // フロントエンドの静的ファイルをホストするためのS3バケットを作成
    const cdkTemplateFrontendBucket = new Bucket(
      this,
      "CdkTemplateFrontendBucket",
      {
        websiteIndexDocument: "index.html",
        publicReadAccess: true,
        autoDeleteObjects: true,
        blockPublicAccess: BlockPublicAccess.BLOCK_ACLS,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        serverAccessLogsBucket: cdkTemplateLogBucket,
        serverAccessLogsPrefix: "frontend-bucket-access-logs/",
      }
    )

    // S3バケットへのアクセスを許可するIAMポリシーステートメントを作成
    const policyStatement = new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["s3:GetObject"],
      principals: [new AnyPrincipal()],
      resources: [cdkTemplateFrontendBucket.arnForObjects("*")],
    })
    cdkTemplateFrontendBucket.addToResourcePolicy(policyStatement)

    // CloudFrontからS3バケットへのアクセスを許可するためのOAIを作成
    const cdkTemplateOAI = new OriginAccessIdentity(
      this,
      "CdkTemplateFrontendOAI"
    )
    cdkTemplateFrontendBucket.grantRead(cdkTemplateOAI)

    // CloudFrontにリクエストが来た際にIP制限を行うCloudFront Functionを作成
    const ipRestrictionFunction = new cdk.aws_cloudfront.Function(
      this,
      "ipRestrictionFunction",
      {
        code: FunctionCode.fromInline(
          readFileSync("./lambda/ip-restriction.js", "utf8")
        ),
        runtime: FunctionRuntime.JS_2_0,
      }
    )

    // CloudFrontディストリビューションを作成し、上記で作成したS3バケットとFunctionを関連付ける
    const distribution = new CloudFrontWebDistribution(
      this,
      "CdkTemplateFrontendWebDestribution",
      {
        originConfigs: [
          {
            s3OriginSource: {
              s3BucketSource: cdkTemplateFrontendBucket,
              originAccessIdentity: cdkTemplateOAI,
            },
            behaviors: [
              {
                isDefaultBehavior: true,
                functionAssociations: [
                  {
                    eventType: FunctionEventType.VIEWER_REQUEST,
                    function: ipRestrictionFunction,
                  },
                ],
              },
            ],
          },
        ],
        // 案件によってログ出力先を適宜変更
        loggingConfig: {
          bucket: cdkTemplateLogBucket,
          prefix: "cloudfront/front",
        },
      }
    )

    // CloudFrontディストリビューションのドメイン名を出力
    new cdk.CfnOutput(this, "CdkTemplateFrontendWebDestributionName", {
      value: distribution.distributionDomainName,
    })

    // Lambda関数のIAMロールを作成
    const lambdaRole = new Role(this, "LambdaExecutionRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        cdk.aws_iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole"
        ),
      ],
    })

    // Lambda関数を作成
    const lambda = new NodejsFunction(this, "TestTemplateLambda", {
      entry: "lambda/index.ts",
      handler: "handler",
      runtime: Runtime.NODEJS_20_X,
      role: lambdaRole,
      logRetention: cdk.aws_logs.RetentionDays.ONE_WEEK,
    })
    // API Gatewayの設定
    const api = new cdk.aws_apigateway.RestApi(this, "LambdaApi", {
      restApiName: "Lambda Service",
      defaultCorsPreflightOptions: {
        allowOrigins: cdk.aws_apigateway.Cors.ALL_ORIGINS,
        allowMethods: cdk.aws_apigateway.Cors.ALL_METHODS,
      },
    })

    // Lambda統合
    const getLambdaIntegration = new cdk.aws_apigateway.LambdaIntegration(
      lambda
    )
    api.root.addMethod("GET", getLambdaIntegration)

    // CloudFrontディストリビューションを介してフロントエンドの静的ファイルをデプロイする
    new BucketDeployment(this, "CdkTemplateFrontendBucketDeployment", {
      sources: [
        Source.asset(path.resolve(__dirname, "../frontend/out")),
        Source.jsonData("env.json", { apigatewayUrl: api.url }),
      ],
      destinationBucket: cdkTemplateFrontendBucket,
      distribution: distribution,
      distributionPaths: ["/*"],
    })

    // API Gatewayエンドポイントを出力
    new cdk.CfnOutput(this, "APIGatewayUrl", {
      value: api.url,
    })
  }
}
