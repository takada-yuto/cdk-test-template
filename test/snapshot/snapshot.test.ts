import * as cdk from "aws-cdk-lib"
import { Template } from "aws-cdk-lib/assertions"
import { CdkTestTemplateStack } from "../../lib/cdk-test-template-stack"

// example test. To run these tests, uncomment this file along with the
// example resource in lib/cdk-test-template-stack.ts
test("snapshot test", () => {
  const app = new cdk.App()
  // WHEN
  const stack = new CdkTestTemplateStack(app, "MyTestStack")
  // THEN
  const template = Template.fromStack(stack)

  expect(template).toMatchSnapshot()
})
