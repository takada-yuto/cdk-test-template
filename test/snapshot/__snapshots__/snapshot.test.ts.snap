// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Outputs": {
    "CdkTemplateFrontendWebDestributionName": {
      "Value": {
        "Fn::GetAtt": [
          "CdkTemplateFrontendWebDestributionCFDistribution5CD766E9",
          "DomainName",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CdkTemplateFrontendBucket12D940AB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "CdkTemplateLogBucket3988B7AB",
          },
          "LogFilePrefix": "frontend-bucket-access-logs/",
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "IgnorePublicAcls": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
          {
            "Key": "aws-cdk:cr-owned:b72ce4b5",
            "Value": "true",
          },
        ],
        "WebsiteConfiguration": {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkTemplateFrontendBucketAutoDeleteObjectsCustomResource52640B91": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CdkTemplateFrontendBucketPolicy211351ED",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "CdkTemplateFrontendBucket12D940AB",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkTemplateFrontendBucketDeploymentAwsCliLayer784A9F99": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "CdkTemplateFrontendBucketDeploymentCustomResource4F500D73": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "CdkTemplateFrontendBucket12D940AB",
        },
        "DistributionId": {
          "Ref": "CdkTemplateFrontendWebDestributionCFDistribution5CD766E9",
        },
        "DistributionPaths": [
          "/*",
        ],
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "HASH-REPLACED.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkTemplateFrontendBucketPolicy211351ED": {
      "Properties": {
        "Bucket": {
          "Ref": "CdkTemplateFrontendBucket12D940AB",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "CdkTemplateFrontendBucket12D940AB",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CdkTemplateFrontendBucket12D940AB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CdkTemplateFrontendBucket12D940AB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "CdkTemplateFrontendOAIBA331240",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CdkTemplateFrontendBucket12D940AB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CdkTemplateFrontendBucket12D940AB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "CdkTemplateFrontendOAIBA331240",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "CdkTemplateFrontendBucket12D940AB",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CdkTemplateFrontendOAIBA331240": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "CdkTemplateFrontendWebDestributionCFDistribution5CD766E9": {
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
            ],
            "CachedMethods": [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "FunctionAssociations": [
              {
                "EventType": "viewer-request",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "ipRestrictionFunctionCBD82F32",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": {
              "Fn::GetAtt": [
                "CdkTemplateLogBucket3988B7AB",
                "RegionalDomainName",
              ],
            },
            "IncludeCookies": false,
            "Prefix": "cloudfront/front",
          },
          "Origins": [
            {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": {
                "Fn::GetAtt": [
                  "CdkTemplateFrontendBucket12D940AB",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "CdkTemplateFrontendOAIBA331240",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "CdkTemplateLogBucket3988B7AB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkTemplateLogBucketAutoDeleteObjectsCustomResource1FA96488": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CdkTemplateLogBucketPolicyF3EFC2A0",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "CdkTemplateLogBucket3988B7AB",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "CdkTemplateLogBucketPolicyF3EFC2A0": {
      "Properties": {
        "Bucket": {
          "Ref": "CdkTemplateLogBucket3988B7AB",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CdkTemplateLogBucket3988B7AB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CdkTemplateLogBucket3988B7AB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "CdkTemplateFrontendBucketDeploymentAwsCliLayer784A9F99",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CdkTemplateFrontendBucket12D940AB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CdkTemplateFrontendBucket12D940AB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "CdkTemplateLogBucket3988B7AB",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ipRestrictionFunctionCBD82F32": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "function handler(event) {
  var request = event.request
  var clientIP = event.viewer.ip
  // 社内IP
  var IP_WHITE_LIST = [
    "122.217.101.122",
    "92.203.96.14",
    "210.203.244.214",
    "92.203.98.151",
    "118.238.205.130",
    "182.171.70.254",
    "182.171.70.250",
    "182.171.70.251",
    "182.171.70.252",
    "182.171.70.253",
    "221.241.133.82",
    "221.241.133.83",
    "221.241.133.84",
    "221.241.133.85",
    "221.241.133.86",
    "210.138.216.235",
    "122.218.252.106",
    "118.238.205.206",
    "183.177.131.6",
    "113.43.254.162",
    "118.238.210.123",
    "113.43.255.130",
    "113.43.255.131",
    "113.43.255.132",
    "39.110.216.150",
    "150.249.255.42",
    "150.249.255.43",
    "124.36.29.42",
    "150.249.255.18",
    "150.249.255.19",
    "124.36.29.46",
    "160.86.110.18",
    "160.86.110.17",
    "122.209.253.250",
    "101.143.175.230",
    "119.243.84.217",
    "219.75.141.252",
    "61.205.12.42",
    "122.213.189.106",
    "113.43.235.134",
    "219.122.228.234",
    "119.243.84.254",
    "203.135.231.180",
    "92.203.98.149",
    "92.203.116.26",
    "92.203.116.30",
    "203.135.194.207",
    "203.135.194.202",
    "61.7.121.49",
    "92.203.98.148",
    "92.203.98.155",
    "183.177.131.203",
    "92.203.116.18",
    "92.203.116.21",
    "61.7.121.44",
    "61.7.121.46",
    "153.246.32.106",
    "153.246.32.107",
    "59.158.15.42",
    "116.82.243.154",
    "152.165.121.149",
    "92.203.98.147",
    "152.165.121.190",
  ]
  var isPermittedIp = IP_WHITE_LIST.includes(clientIP)

  if (isPermittedIp) {
    return request
  }
  return {
    statusCode: 403,
    statusDescription: "Forbidden",
  }
}
",
        "FunctionConfig": {
          "Comment": {
            "Fn::Join": [
              "",
              [
                {
                  "Ref": "AWS::Region",
                },
                "MyTestStackipRestrictionFunction1E404FAE",
              ],
            ],
          },
          "Runtime": "cloudfront-js-2.0",
        },
        "Name": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::Region",
              },
              "MyTestStackipRestrictionFunction1E404FAE",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFront::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
